### ISSUE TITLE
"import Redis" is flagged as invalid constructor function type in ESM TypeScript project

### ISSUE THREAD

**Issue by Sayakie** [luin/ioredis#1642]
I had to do this to get started
```ts
import Redis from 'ioredis'

import type { Species } from '#lib/pokemon/Species.js'

export enum RedisKey {
  GetPokemon = `getPokemon`
}

export type RedisQuery<K extends RedisKey> = K extends `getPokemon` ? Species : null

export class RedisClient extends Redis {
  public async insert<K extends RedisKey>(
    key: K,
    query: RedisQuery<K>,
    data: unknown
  ): Promise<`OK`> {
    return super.set(`${key}:${query}`, JSON.stringify(data))
  }
}
```

Same issue as:
```ts
import { default as Redis } from 'ioredis'
```

They prints: `Type 'typeof import("node_modules/.pnpm/ioredis@5.2.3/node_modules/ioredis/built/index")' is not a constructor function type.`

My project settings are:

|  package        |  version       |
| --------------- | ---------------|
| Module Type     | ESM            |
| Node.js         | 16.13.2        |
| pnpm            | 7.9.5          |
| ioredis         | 5.2.3          |
| TypeScript      | 4.8.2          |
| @types/ioredis  | not installed  |

TSConfig:
```json
{
  "extends": "@sapphire/ts-config/extra-strict",
  "compilerOptions": {
    "lib": ["ESNext"],
    "target": "ESNext",
    "module": "ESNext",
    "moduleResolution": "Node16",
    "emitDecoratorMetadata": false
  }
}
```

**Comment by Sayakie** [comment]
Additional info for more context, `new Redis()` throws similar error: "`This expression is not constructable. Type 'typeof import("node_modules/.pnpm/ioredis@5.2.3/node_modules/ioredis/built/index")' has no construct signatures.`"

![Append "new Redis()" to initialize ioredis class, but it throws "This expression is not constructable." in fly-in-compile-time on Visual Studio Code](https://user-images.githubusercontent.com/18323202/187428633-1637cdb7-eb52-4902-8266-f28c18147c5f.png)

**Comment by Sayakie** [comment]
Resolved as do:
```ts
import { default as Redis } from 'ioredis'

export class RedisClient extends Redis.default {
  // ... whatever
}
```

**Comment by GingerAdonis** [comment]
I'd say this issue should be re-opened as the workaround is kind of odd.

**Comment by Sayakie** [comment]
> I'd say this issue should be re-opened as the workaround is kind of odd.

I assume would be better open a new ticket.

**Comment by GingerAdonis** [comment]
> > I'd say this issue should be re-opened as the workaround is kind of odd.
> 
> I assume would be better open a new ticket.
I disagree, as it would be a duplicate of this unresolved issue. But I'll do that if you don't want to open the issue.

**Comment by Sayakie** [comment]
@luin I re-open this issue because the behavior is odd and still would be problem.

Tracking of #1624, #1632

**Comment by luin** [comment]
`import { default as Redis } from "ioredis"` is the solution documented on README.md but I agree the syntax may seem odd.

**Comment by GingerAdonis** [comment]
> `import { default as Redis } from "ioredis"` is the solution documented on README.md but I agree the syntax may seem odd.

The problem is that despite using that import statement one still has to use `Redis.default` as a constructor to make it properly work with Typescript.

Example:
```ts
import { default as Redis } from 'ioredis';
const redis = new Redis.default();
```

The named Redis export is only exported as a type. Which is kind of strange as well.

**Comment by LinusU** [comment]
I can confirm that the normal import doesn't work when using `moduleResolution` `node16` or `nodenext`. We are currently using the following workaround:

```ts
import RedisModule from 'ioredis'

// FIXME: https://github.com/luin/ioredis/issues/1642
const Redis = RedisModule.default
```

**Comment by huangyingjie** [comment]
> I can confirm that the normal import doesn't work when using `moduleResolution` `node16` or `nodenext`. We are currently using the following workaround:
> 
> ```ts
> import RedisModule from 'ioredis'
> 
> // FIXME: https://github.com/luin/ioredis/issues/1642
> const Redis = RedisModule.default
> ```

I also use node16, and trapped by this @luin

**Comment by huangyingjie** [comment]
> I can confirm that the normal import doesn't work when using `moduleResolution` `node16` or `nodenext`. We are currently using the following workaround:
> 
> ```ts
> import RedisModule from 'ioredis'
> 
> // FIXME: https://github.com/luin/ioredis/issues/1642
> const Redis = RedisModule.default
> ```

I also use node16, and trapped by this @luin

**Comment by huangyingjie** [comment]
> I can confirm that the normal import doesn't work when using `moduleResolution` `node16` or `nodenext`. We are currently using the following workaround:
> 
> ```ts
> import RedisModule from 'ioredis'
> 
> // FIXME: https://github.com/luin/ioredis/issues/1642
> const Redis = RedisModule.default
> ```

I also use node16, and trapped by this @luin

**Comment by GingerAdonis** [comment]
This issue has been fixed. Probably in [v5.2.5](https://github.com/luin/ioredis/releases/tag/v5.2.5).

Issue can be closed.

**Comment by Sayakie** [comment]
Fixed in #1695

**Comment by sinisarudan** [comment]
I'm having this issue with
"ioredis": "^5.5.0",

tsconfig.json:

```json
{
  "compilerOptions": {
    "target": "ES2021",
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "declaration": true,
    "removeComments": true,
    "resolveJsonModule": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "allowSyntheticDefaultImports": true,
    "sourceMap": true,
    "outDir": "./dist",
    "rootDir": "./",
    "baseUrl": "./",
    "paths": {
      "@app/*": ["src/app/*"]
    },
    "incremental": true,
    "strict": true,
    "skipLibCheck": true,
    "strictNullChecks": true,
    "noImplicitAny": false,
    "strictBindCallApply": true,
    "forceConsistentCasingInFileNames": true,
    "noFallthroughCasesInSwitch": true,
    "esModuleInterop": true,
    "moduleDetection": "force",
    "isolatedModules": true,
    "allowJs": false,
    "types": ["node"]
  },
  "include": ["src/**/*.ts", "index.ts"],
  "exclude": ["node_modules", "dist", "uploads"],
  "files": ["./src/app/api/api-types.d.ts"],
}
```
